omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/DiscreteValuesFragmentController.java

```diff
@@ -0,0 +1,143 @@
+package org.openmrs.module.pihcore.fragment.controller.dashboardwidgets;
+
+import org.apache.commons.lang.StringUtils;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.node.ArrayNode;
+import org.openmrs.Concept;
+import org.openmrs.Encounter;
+import org.openmrs.Obs;
+import org.openmrs.Patient;
+import org.openmrs.PatientProgram;
+import org.openmrs.Program;
+import org.openmrs.api.ConceptService;
+import org.openmrs.api.ObsService;
+import org.openmrs.api.ProgramWorkflowService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.appframework.domain.AppDescriptor;
+import org.openmrs.module.emrapi.patient.PatientDomainWrapper;
+import org.openmrs.ui.framework.SimpleObject;
+import org.openmrs.ui.framework.annotation.FragmentParam;
+import org.openmrs.ui.framework.annotation.SpringBean;
+import org.openmrs.ui.framework.fragment.FragmentConfiguration;
+import org.openmrs.ui.framework.fragment.FragmentModel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DiscreteValuesFragmentController {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public void controller(
+                           @SpringBean("conceptService") ConceptService conceptService,
+                           @SpringBean("obsService") ObsService obsService,
+                           @SpringBean("programWorkflowService") ProgramWorkflowService programWorkflowService,
+                           @FragmentParam("app") AppDescriptor app,
+                           FragmentConfiguration config,
+                           FragmentModel model) {
+
+        Object patientConfig = config.get("patient");
+        Patient patient = null;
+        if (patientConfig != null && (patientConfig instanceof PatientDomainWrapper)) {
+            patient = ((PatientDomainWrapper) patientConfig).getPatient();
+        }
+        if (patient == null) {
+            throw new IllegalArgumentException("No patient found. Please pass patient into the configuration");
+        }
+
+        model.put("app", app);
+        model.put("patient", patient);
+
+        JsonNode node = app.getConfig().get("fields");
```

Comment:
I think this same issue probably exists everywhere / all over the place, but this is something I was contending with in other fragments I was recently writing for Rwanda.  Basically, I think all of the config that drives the behavior of the fragment should be pulled from the `FragmentConfiguration config`, _not_ from the `app.getConfig()`.  The app config is for configuring the "app", whereas the fragment config should be for configuring the fragment.

Think of it this way - if you wanted to add this same fragment to a gsp directly (eg. without adding it as an app/extension), could you do it?  If the configuration it needs is coming from an "app.getConfig()" then you couldn't.  But if it is coming from the fragment config, then you simply put the configuration into the fragment declaration on the page, and it works as expected.

Does this make sense?  The same applies to any configurations here

Reply:
@mseaton , makes sense. How would you do this? Is this fragment config configured this way anywhere else? I could not find an example on how to change it. Thanks!

Reply:
@mseaton , is this what you meant? using fragmentConfig param?
![Screenshot 2024-10-10 at 2 08 10 PM](https://github.com/user-attachments/assets/cf73ce94-2431-4f57-9a20-57c0281c4fda)

Reply:
fyi @mseaton , the above, using the fragmentConfig worked fine. I will refactor this widget to use the fragmentConfig node.

Reply:
Yes, that's right @cioan